### ISSUE TITLE
Large timestamps (9999-12-31) not read correctly

### ISSUE THREAD

**Issue by george-zubrienko** [delta-io/delta-rs#578]
# Environment

**Delta-rs version**: 0.5.6

**Binding**: Python (deltalake 0.5.6)

**Environment**:
- **Cloud provider**: Azure
- **OS**: Ubuntu 20.04
- **Other**: Python 3.9 (poetry venv)

***
# Bug

**What happened**:

The following code fails when trying to read a delta table that has TIMESTAMP column with `9999-12-31 00:00:00` values instead of nulls:
```
    delta_table = DeltaTable(path.to_delta_rs_path(), version=version)

    if batch_size:
        batches: Iterator[RecordBatch] = delta_table.to_pyarrow_dataset().to_batches(filter=row_filter, columns=columns, batch_size=batch_size)

        return map(lambda batch: batch.to_pandas(), batches)

    pyarrow_table: Table = delta_table.to_pyarrow_dataset().to_table(filter=row_filter, columns=columns)

    result = pyarrow_table.to_pandas()
```

Trace:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/c/users/gzu/source/repos/github/proteus/proteus/storage/delta_lake/_functions.py", line 83, in load
    pyarrow_table: Table = delta_table.to_pyarrow_dataset().to_table(
  File "pyarrow/_dataset.pyx", line 323, in pyarrow._dataset.Dataset.to_table
  File "pyarrow/_dataset.pyx", line 2311, in pyarrow._dataset.Scanner.to_table
  File "pyarrow/error.pxi", line 143, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 99, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Casting from timestamp[ns] to timestamp[us] would lose data: -4852202631933722624
```

If `columns` doesn't contain a column that has `9999-12-31 00:00:00` timestamp values, everything works fine.

**What you expected to happen**: `result` contains pandas DataFrame and doesn't throw.

**How to reproduce it**:

Create a delta table with a timestamp column that contains `9999-12-31 00:00:00` values (from Spark, so it is stored as INT96).

**More details**:
I was able to fix this issue by changing `to_pyarrow_dataset` method:

```
def _to_pyarrow_dataset(
        table: RawDeltaTable,
        schema: pyarrow.Schema,
        partitions: Optional[List[Tuple[str, str, Any]]] = None,
        filesystem: Optional[Union[str, pa_fs.FileSystem]] = None,
) -> pyarrow.dataset.Dataset:
    """
    Build a PyArrow Dataset using data from the DeltaTable.

    :param partitions: A list of partition filters, see help(DeltaTable.files_by_partitions) for filter syntax
    :param filesystem: A concrete implementation of the Pyarrow FileSystem or a fsspec-compatible interface. If None, the first file path will be used to determine the right FileSystem
    :return: the PyArrow dataset in PyArrow
    """
    if not filesystem:
        filesystem = pa_fs.PyFileSystem(
            DeltaStorageHandler(table.table_uri())
        )

    format = ParquetFileFormat(read_options=ParquetReadOptions(coerce_int96_timestamp_unit='ms'))

    fragments = [
        format.make_fragment(
            file,
            filesystem=filesystem,
            partition_expression=part_expression,
        )
        for file, part_expression in table.dataset_partitions(partitions)
    ]

    return FileSystemDataset(fragments, schema, format, filesystem)
```

Specifically, doing `format = ParquetFileFormat(read_options=ParquetReadOptions(coerce_int96_timestamp_unit='ms'))` instead of source code having `format = ParquetFileFormat()`

Then the pyarrow conversion will go through, but pandas conversion will fail. It is fixed via `pyarrow_table.to_pandas(timestamp_as_object=True)` instead of `pyarrow_table.to_pandas()`

## TLDR

Parquet read options are not exposed in public API. Would be nice to have them there, so issues like these can be resolved automatically :)

**Comment by wjones127** [comment]
Thanks for the report! I think we want to be compatible with Spark by default, so hard-coding `ParquetReadOptions(coerce_int96_timestamp_unit='ms'))` seems appropriate to me.

But I also agree we should allow passing through format and writer kwargs as well. For writer args, there are options controlling sizes of row groups and files that users might want to tune.

Feel free to submit a PR for this if you want, or else I will look at fixing this soon.

**Comment by george-zubrienko** [comment]
I think it is enough just to expose parquet read/write options - it would be nice to be able to control row group size when writing as you mentioned. I'll try to find some time and send a proposal PR this weekend.

**Comment by george-zubrienko** [comment]
@wjones127 please take a look at this [draft](https://github.com/delta-io/delta-rs/pull/581). If/when the exposed options are agreed, I'll add unit tests for the corner case mentioned in this issue, as well as 0001-01-01 date
